src/GovUk.Education.ExploreEducationStatistics.Common/Services/PostgreSqlRepository.cs

```diff
@@ -0,0 +1,99 @@
+        CancellationToken cancellationToken = default)
+        where TDbContext : DbContext
+        where TResponse : class
+    {
+        var jsonPathParam = new NpgsqlParameter("jsonPath", request.PathSegments);
+        var rowIdParam = new NpgsqlParameter("rowId", request.RowId);
+
+        var response = await context
+            .Set<JsonFragment>()
+            .FromSqlRaw(
+                sql: $"""
+                      SELECT "{request.JsonbColumnName}"#>@jsonPath "JsonString" FROM "{request.TableName}"
+                      WHERE "{request.IdColumnName}" = @rowId
+                      """,
+                parameters: [jsonPathParam, rowIdParam])
+            .SingleOrDefaultAsync(cancellationToken);
+
+        return response is not null
+               && response.JsonString is not null
+            ? JsonSerializer.Deserialize<TResponse>(response.JsonString)
+            : null;
+    }
+
+#pragma warning disable EF1002
+    public async Task<TValue?> SetJsonbAtPath<TDbContext, TValue, TRowId>(
+        TDbContext context,
+        JsonbPathRequest<TRowId> request,
+        TValue? value,
+        CancellationToken cancellationToken = default)
+        where TDbContext : DbContext
+        where TValue : class
+    {
+        var jsonPathParam = new NpgsqlParameter("jsonPath", request.PathSegments);
+        var rowIdParam = new NpgsqlParameter("rowId", request.RowId);
+        var valueParam = new NpgsqlParameter("value", NpgsqlDbType.Json)
+        {
+            Value = value is not null
+                ? JsonSerializer.Serialize(value)
+                : null
+        };
+
+        await context
+            .Database
+            .ExecuteSqlRawAsync(
+                sql: $"""
+                      UPDATE "{request.TableName}" SET "{request.JsonbColumnName}" =
+                      jsonb_set("{request.JsonbColumnName}", @jsonPath, to_jsonb(@value))
+                      WHERE "{request.IdColumnName}" = @rowId
+                      """,
+                parameters: [jsonPathParam, valueParam, rowIdParam],
+                cancellationToken: cancellationToken);
+
+        return value;
+    }
+
+#pragma warning restore EF1002
+
+    public async Task<Either<TFailure, TValue?>> UpdateJsonbAtPath<TDbContext, TValue, TRowId, TFailure>(
+        TDbContext context,
+        JsonbPathRequest<TRowId> request,
+        Func<TValue?, Either<TFailure, TValue?>> updateValueFn,
+        CancellationToken cancellationToken = default)
+        where TDbContext : DbContext
+        where TValue : class
+    {
+        var json = await GetJsonbFromPath<TDbContext, TRowId, TValue>(
+            context,
+            request,
+            cancellationToken);
+
+        var jsonUpdateResult = updateValueFn.Invoke(json);
+
+        if (jsonUpdateResult.IsLeft)
+        {
+            return jsonUpdateResult;
+        }
+
+        return await SetJsonbAtPath(context, request, jsonUpdateResult.Right, cancellationToken);
```

Comment:
I think you could simplify this to:

```csharp
        return updateValueFn.Invoke(json)
            .OnSuccess(async jsonUpdateResult => await SetJsonbAtPath(context, request, jsonUpdateResult.Right, cancellationToken));
```

Reply:
I can if I cheat a bit and have `updateValueFn` return a `Task<Either>` rather than a straight `Either` :grin: 

And I'm actually happy to make this change despite the fact that we're not currently needing Task support in our current use case, because this makes it more flexible for the future.

So I've updated the method signature and added this simplfication too, thanks!

Reply:
I thought we had an `.OnSuccess` extension for use with non-task `Eithers`? So I thought you might have been ok to leave the `updateValueFn` type as if was